pkg/restore/cdclog/buffer.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdclog
+
+import (
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/types"
+	"go.uber.org/zap"
+)
+
+// TableBuffer represents the kv buffer of this table.
+// we restore one tableBuffer in one goroutine.
+// this is the concurrent unit of log restore.
+type TableBuffer struct {
+	KvPairs []Row
+	Size    int64
+
+	KvEncoderFn func(autoid.Allocators, table.Table) (Encoder, error)
+	KvEncoder   Encoder
+	tableInfo   table.Table
+	allocator   autoid.Allocators
+
+	flushKVPairs int
+
+	colNames []string
+	colPerm  []int
+}
+
+// NewTableBuffer creates TableBuffer.
+func NewTableBuffer(tbl table.Table, allocators autoid.Allocators, flushKVPairs int) *TableBuffer {
+	kvEncoderFn := func(allocators autoid.Allocators, tbl table.Table) (Encoder, error) {
+		encTable, err := table.TableFromMeta(allocators, tbl.Meta())
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return NewTableKVEncoder(encTable, &SessionOptions{
+			Timestamp: time.Now().Unix(),
+			// TODO get the version from TiDB cluster
+			// currently TiDB only support v1 and v2, and since 4.0
+			// the default RowFormatVersion is 2, so I think
+			// we can implement the row version retrieve from cluster in the future
+			// when TiDB decide to support v3 RowFormatVersion.
+			RowFormatVersion: "2",
```

Comment:
why not directly read this info from tidb cluster like lightning do?

Reply:
I explain it in above comment.

Reply:
Currently we don't have tidb mgr like lightning in BR. BR use tidb.Domain to interact with TiKV. it's not easy to implement it in BR like lightning.